lib/auth.js

```diff
@@ -1067,6 +1066,164 @@ function authorize(req, res, next) {
+ *                              value here even if you get a
+ *                              UserDoesNotExistError from mahi.getUser().)
+ *
+ *             key              The internal resource path (i.e., with the
+ *             (string)         user's login resolved to their user uuid).  This
+ *                              is used to check for access to objects in
+ *                              directories that are defined to be public (e.g.,
+ *                              "/$userid/public/...").
+ *
+ *             roles            Role tags (uuids) associated with this resource.
+ *             (array)
+ *
+ *          conditions  Additional context collected as part of this operation:
+ *          (object)
+ *
+ *              activeRoles     List of role tags that are active for the
+ *              (array)         current user for this operation.
+ *
+ *              owner           Represents the *account* that owns the resource.
+ *              (object)        This differs from resource.owner, which may
+ *                              actually specify the anonymous subuser (see
+ *                              above).  It's useful to remember that
+ *                              "conditions" are used to evaluate user-specified
+ *                              access control rules, so "owner" exists in this
+ *                              context to allow users to write rules based on
+ *                              the account that owns the resource.  By
+ *                              contrast, "owner" exists under "resource" for
+ *                              the system-defined semantics relating to
+ *                              a resource's owner.
+ */
+function _authorize(opts) {
+    assert.object(opts);
+    assert.object(opts.mahi, 'mahi');
+    assert.object(opts.context, 'context');
+    assert.object(opts.context.principal, 'principal');
+    assert.string(opts.context.action, 'action');
+    assert.object(opts.context.resource, 'resource');
+    assert.object(opts.context.conditions, 'conditions');
+
+    assert.object(opts.context.principal.account, 'principal.account');
+    assert.optionalObject(opts.context.principal.user, 'principal.user');
+
+    assert.object(opts.context.resource.owner, 'resource.owner');
+    assert.string(opts.context.resource.owner.account.uuid,
+        'resource.owner.account.uuid');
+    assert.arrayOfString(opts.context.resource.roles, 'resource.roles');
+    assert.string(opts.context.resource.key, 'resource.key');
+
+    assert.string(opts.context.conditions.method, 'conditions.method');
+    assert.arrayOfString(opts.context.conditions.activeRoles,
+        'condition.activeRoles'); // array of UUID
+
+    var mahi = opts.mahi;
+    var action = opts.context.action;
+    var resource = opts.context.resource;
+    resource.roles = resource.roles || [];
+
+    // Authorize if public GET.
+    if (common.PUBLIC_STOR_PATH.test(resource.key) &&
```

Comment:
we can probably remove this for buckets, right?

Reply:
yeah